ferry.go

```diff
@@ -1021,3 +1030,84 @@ func (f *Ferry) checkConnectionForBinlogFormat(db *sql.DB) error {
 
 	return nil
 }
+
+func (f *Ferry) UpdateBatchSizes() error {
+	schemaTablesMap := map[string][]string{}
+
+	for _, table := range f.Tables {
+		if _, found := f.Config.DataIterationBatchSizePerTableOverride.TableOverride[table.Schema][table.Name]; found {
+			continue
+		}
+		schemaTablesMap[table.Schema] = append(schemaTablesMap[table.Schema], table.Name)
+	}
+
+	for schemaName, tables := range schemaTablesMap {
+		if _, found := f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName]; !found {
+			f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName] = map[string]uint64{}
+		}
+		query := fmt.Sprintf(
+			`SELECT T1.TABLE_NAME, AVG_ROW_LENGTH, MAX_BATCH_SIZE
+					FROM information_schema.TABLES T1
+						JOIN (SELECT TABLE_NAME, TABLE_SCHEMA, floor(%d / count(*)) MAX_BATCH_SIZE
+						FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = '%s' AND TABLE_NAME in (%s)
+						GROUP BY table_name) T2 USING (TABLE_SCHEMA, TABLE_NAME);`,
+			MySQLNumParamsLimit,
+			schemaName,
+			"'"+strings.Join(tables, "', '")+"'",
+		)
+		rows, err := f.SourceDB.Query(query)
+		if err != nil {
+			return err
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var avgRowLength, maxRowLength int
+			var tableName string
+			err = rows.Scan(&tableName, &avgRowLength, &maxRowLength)
+			if err != nil {
+				return err
+			}
+			f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName][tableName] = f.calculateBatchSize(avgRowLength, maxRowLength)
+		}
+
+	}
+	return nil
+}
+
+func (f *Ferry) calculateBatchSize(rowLength, maxRowLength int) uint64 {
+	batchSizePoints := f.Config.DataIterationBatchSizePerTableOverride
+
+	if batchSize, found := batchSizePoints.ControlPoints[rowLength]; found {
+		return batchSize
+	}
+	if rowLength >= batchSizePoints.MaxAvgRowLength {
+		return batchSizePoints.ControlPoints[batchSizePoints.MaxAvgRowLength]
+	}
+	if rowLength <= batchSizePoints.MinAvgRowLength {
+		return batchSizePoints.ControlPoints[batchSizePoints.MinAvgRowLength]
+	}
+	if rowLength >= maxRowLength {
```

Comment:
need to move this above L1083 because database "max" should override anything we write in the config.

Reply:
This is my mistake, we should actually be checking the result of `calculateBatchSize()` against `MaxBatchSize` not using `avgRowLength`

updated to check after we return from `calculateBatchSize()`
```
batchSize := f.calculateBatchSize(avgRowLength)
if batchSize >= maxBatchSize{
  batchSize = maxBatchSize
}
f.Config.DataIterationBatchSizePerTableOverride.TableOverride[schemaName][tableName] = uint64(batchSize)
```